### PR TITLE
Feature(ConnectorSdk): PriorityFirstSync Example

### PR DESCRIPTION
**Pull request by fivetran-chiragramachandraiah**
Links https://fivetran.height.app/T-767058

PFS Example

**Testing**
Tested locally. Both incremental and historical syncs work fine.

### CODE REVIEW

**Review by varundhall — CHANGES_REQUESTED** [review]
please make the changes as discussed.

**Comment by fivetran-chiragramachandraiah** [comment]
Have made the changes. It's ready for review.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: please follow our existing examples for standard links to docs etc

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]:
```suggestion
SYNC_DURATION_THRESHOLD_IN_HOURS = 6
```

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]:
```suggestion
HISTORICAL_SYNC_LIMIT = 'historical_limit'
```

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]:
```suggestion
INCREMENTAL_SYNC_CURSOR = 'incremental_cursor'
```

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: add a comment what this base url is about, as you already have mock API, customer may get confused with fivetran-api-playground.

Better to use something like `example.com/xyz`

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: add docs/comments for all the methods

Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]:
```suggestion
def is_historical_data_completely_synced(state, endpoints):
```
this is just a suggestion, you can tweak it

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]:
```suggestion
def run_incremental_sync_for_endpoints(state, endpoints):
```

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py:24]: I'm thinking do we even need this, can we just simplify the cursors to incremental and historical

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]:
```suggestion
def get_api_response(base_url, params):
```
mock is already written in file name, no need to mention it twice

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]:
```suggestion
    if until > datetime.now(timezone.utc):
```

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]: nit: you can use faker https://pypi.org/project/Faker/
and mention it in requirements.txt to generate better data

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Hey @varundhall I feel method names are descriptive, I don't think comments will help make it more descriptive. Lmk particular methods instead you feel comments are needed. It's possible customers may find too many code comments to be distracting.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py:24]: Currently code is generic and supports multiple endpoints although only single endpoint is used in example. Are you suggesting to strip current generic code and write code in such a way as if pfs supports only single endpoint? Like we will not have the for loop for endpoints, we will directly call the endpoint method and pfs will be implemented for that single endpoint. Just want to confirm before making the change as it a big one. Cc @varundhall

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done

File: README.md
Comment (alsnkline) [inline README.md:72]: I think we should include a link to our docs on the approach:
https://fivetran.com/docs/using-fivetran/features#priorityfirstsync

File: README.md
Comment (alsnkline) [inline README.md]:
```suggestion
which prioritises fetching the most recent data first so that fresh data is ready for you to use more quickly.
```

File: README.md
Comment (alsnkline) [inline README.md]:
```suggestion
This is a simple example of how you could implement the Priority-first sync strategy in a `connector.py` file for your connection.
```

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (alsnkline) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: for the above 3 lines match those in the readme - once they have ben finalized.

File: README.md
Comment (fivetran-satvikpatil) [inline README.md:71]: Please get this reviewed by Tech writers.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (alsnkline) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py:24]: for example purposes if the cursor and code changes significantly we should have 2 examples - one with only 1 table and one with multiple?

Also from the discussion I'm thinking we should made sure this (these) example work with the playground and possibly even recommend customers work through the example with the playground before trying to implement in their own connector/end point to help ensure they fully understand how it works.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]: Thanks for pointing.

**Comment by fivetran-chiragramachandraiah** [comment]
@varundhall Addressed most comments. I've replied on 2. Pls check.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-satvikpatil) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: NIT: reuse `set_pfs_incremental_sync` method here.

**Review by alsnkline — COMMENTED** [review]
looking good

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py
Comment (fivetran-satvikpatil) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py:28]: Can you please explain what does this check do?

File: README.md
Comment (fivetran-chiragramachandraiah) [reply README.md:72]: @varundhall  suggested not to. Replied in slack.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py
Comment (fivetran-satvikpatil) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py]: Historical cursor should go backwards right? But if there are multiple pages then we are assigning `params['updated_since']` with `last_updated_at` which will then represent the latest timestamp, as you are assuming the data is in ascending order?

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (fivetran-satvikpatil) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]: We can have an endpoint for this mock data generator in our playground. Not as part of this ticket, but you can do it later.
As I am using a free API for the user in `user_profiles` examplese, I will be able to reuse the same endpoint there as well.
cc: @varundhall

**Review by fivetran-satvikpatil — COMMENTED** [review]
Added few comments.

File: README.md
Comment (alsnkline) [reply README.md:72]: Ok for now then.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py:28]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py]: Yep fixed it.

File: README.md
Comment (fivetran-chiragramachandraiah) [reply README.md:71]: Is it required? There was no review by tech writers in older prs.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]: Yes will create a follow up ticket.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: I've renamed methods. Pls lmk specific places where you think comments is needed.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done.

File: README.md
Comment (fivetran-chiragramachandraiah) [reply README.md]: Done.

File: README.md
Comment (fivetran-chiragramachandraiah) [reply README.md]: Done.

File: README.md
Comment (fivetran-satvikpatil) [reply README.md:71]: We should actually have it reviewed by tech writers as well, but if Alison is reviewing, it should be fine.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: in days hours mins?

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py:1]: please reformat the file via Code->Reformat File in PyCharm

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]: please reformat the file

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py
Comment (varundhall) [inline examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py:1]: please reformat the file

**Comment by varundhall** [comment]
@fivetran-satvikpatil please check locally once before approving the PR.

**Comment by fivetran-chiragramachandraiah** [comment]
> please address the minor comments before merging,
> also I think you need a __init__.py file, as its a package.

I think I will be skipping. From what I read online, it's not required anymore. It maybe was required for older python versions.

> Please make sure you example works properly by running fivetran debug at least few times.

Sure
Cc: @varundhall

**Comment by fivetran-chiragramachandraiah** [comment]
@varundhall pls release the python library with Rajat's fix. Else it will not work for Satvik. @fivetran-satvikpatil upgrade library once it's released.

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/users_sync.py:1]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/mock_api.py]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py:1]: Done

File: examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py
Comment (fivetran-chiragramachandraiah) [reply examples/common_patterns_for_connectors/priority_first_sync_for_high_volume_initial_syncs/connector.py]: Done

**Comment by rajat-kushwaha** [comment]
> @varundhall pls release the python library with Rajat's fix. Else it will not work for Satvik. @fivetran-satvikpatil upgrade library once it's released.

This has now been released.

**Comment by fivetran-satvikpatil** [comment]
@fivetran-chiragramachandraiah While running `fivetran debug` the example is failing with an error (attaching screenshot), please resolve it. I will approve it once its done.
Screenshot 2024-10-10 at 11 51 48 AM

**Comment by fivetran-satvikpatil** [comment]
When we run `fivetran debug`, we don't install the library using `requirements.txt`. So it was throwing me the error regarding this library. I installed it manually, it is working fine.

**Review by fivetran-satvikpatil — APPROVED** [review]

**Pull request merged**
